libraries/api-extractor/src/PackageJsonLookup.ts

```diff
@@ -0,0 +1,76 @@
+/* tslint:disable:no-constant-condition */
+
+import * as fsx from 'fs-extra';
+import * as path from 'path';
+import JsonFile from './JsonFile';
+
+/**
+ * Represents a package.json file.
+ */
+interface IPackageJson {
+  name: string;
+}
+
+/**
+ * This class provides methods for finding the nearest "package.json" for a folder
+ * and retrieving the name of the package.  The results are cached.
+ */
+export default class PackageJsonLookup {
+  private _tryFindPackagePathUpwardsCache: Map<string, string>;
+  private _readPackageNameCache: Map<string, string>;
+
+  constructor() {
+    this._tryFindPackagePathUpwardsCache = new Map<string, string>();
+    this._readPackageNameCache = new Map<string, string>();
+  }
+
+  /**
+   * Finds the path to the package folder of a given currentPath, by probing
+   * upwards from the currentPath until a package.json file is found.
+   * If no package.json can be found, undefined is returned.
+   *
+   * @param currentPath - a path (relative or absolute) of the current location
+   * @returns a relative path to the package folder
+   */
+  public tryFindPackagePathUpwards(currentPath: string): string {
+    // Since Map cannot store an undefined value, we use the weird "null" object to
+    // represent "undefined".
+    let result: string = this._tryFindPackagePathUpwardsCache.get(currentPath);
+    if (result !== undefined) {
+      return result === null ? undefined : result; // tslint:disable-line: no-null-keyword
```

Comment:
@qz2017 actually the code is not the same.  However since that wasn't obvious, I suppose it would be more readable to move it into a function.  :-)

---
In reply to: [122495414](https://github.com/Microsoft/web-build-tools/pull/236#discussion_r122495414) [](ancestors = 122495414)